idea-plugin/src/com/reason/ide/format/ORDocumentFormatter.java

```diff
@@ -0,0 +1,150 @@
+  }
+
+  public ORDocumentFormatter(Project project) {
+    m_project = project;
+    ORFormatCommand oCamlFormatCommand = new OCamlFormatCommand(m_project);
+    RefmtCommand refmtCommand = new RefmtCommand(m_project);
+    m_formatCommands = ImmutableSet.of(
+        Pair.create(FileHelper::isOCaml, oCamlFormatCommand),
+        Pair.create(FileHelper::isReason, refmtCommand),
+        Pair.create(FileHelper::isRescript, refmtCommand)
+    );
+  }
+
+  public void format(@NotNull PsiFile psiFile) {
+    VirtualFile virtualFile = psiFile.getVirtualFile();
+    if (virtualFile.exists() && psiFile.isWritable()) {
+      Optional<ORFormatCommand> command =
+          getCommandForFileType(psiFile.getFileType());
+      command.ifPresent(formatCommand -> formatWithRetry(psiFile, formatCommand));
+    }
+  }
+
+  private void formatWithRetry(@NotNull PsiFile psiFile, ORFormatCommand command) {
+    formatWithRetry(psiFile, command, 0);
+  }
+
+  private void formatWithRetry(@NotNull PsiFile psiFile, ORFormatCommand command, final int retries) {
+    Document document = PsiDocumentManager.getInstance(m_project)
+        .getDocument(psiFile);
+    assert document != null;
+    long before = document.getModificationStamp();
+    String oldText = document.getText();
+    if (!oldText.isEmpty()) {
+      String newText = command.execute(psiFile);
+      if (newText != null && !newText.isEmpty() && !oldText.equals(newText)) { // additional protection
+        ApplicationManager.getApplication()
+            .invokeLater(
+                () -> {
+                  long after = document.getModificationStamp();
+                  if (after > before) {
+                    // Document has changed, redo format one time
+                    if (retries < 2) {
+                      formatWithRetry(psiFile, command, retries + 1);
+                    }
+                  } else {
+                    CommandProcessor.getInstance()
+                        .executeCommand(
+                            m_project,
+                            () -> {
+                              WriteAction.run(
+                                  () -> {
+                                    document.setText(newText);
+                                    FileDocumentManager.getInstance().saveDocument(document);
+                                  });
+                              psiFile.putUserData(UndoConstants.FORCE_RECORD_UNDO, null);
+                            },
+                            "or.reformat",
+                            "CodeFormatGroup",
+                            document);
+                  }
+                });
+      }
+    }
+  }
+
+  private Optional<ORFormatCommand> getCommandForFileType(FileType fileType) {
+    return m_formatCommands.stream()
+        .filter((pair) -> pair.getFirst().test(fileType))
+        .map((pair) -> pair.getSecond())
+        .findFirst();
+  }
+
+  private interface ORFormatCommand {
```

Comment:
maybe this interface could be implemented directly by RefmtProcess & OcamlFormatProcess, and we could get rid of the inner implementations ?